src/core/SamplePreviewPlayHandle.cpp

```diff
@@ -0,0 +1,142 @@
+{
+#ifdef LMMS_BUILD_WIN32
+	m_sndfile = sf_wchar_open(path.c_str(), SFM_READ, &m_sfinfo);
+#else
+	m_sndfile = sf_open(path.c_str(), SFM_READ, &m_sfinfo);
+#endif
+
+	if (m_sndfile == nullptr || m_sfinfo.channels == 0 || m_sfinfo.frames == 0)
+	{
+		throw std::runtime_error{"Failed to create sample preview stream"};
+	}
+
+	const auto initialFrameCount = std::min(static_cast<sf_count_t>(m_sfinfo.samplerate), m_sfinfo.frames);
+	m_buffer.resize((initialFrameCount + 1) * m_sfinfo.channels);
+	sf_readf_float(m_sndfile, m_buffer.data(), initialFrameCount);
+
+	m_frameWriteIndex = (m_frameWriteIndex + initialFrameCount) % framesInBuffer();
+	m_diskStream = ThreadPool::instance().enqueue(&SamplePreviewPlayHandle::runDiskStream, this);
+	setAudioBusHandle(new AudioBusHandle("SamplePreviewPlayHandle", false));
+}
+
+SamplePreviewPlayHandle::~SamplePreviewPlayHandle() noexcept
+{
+	m_quit = true;
+	m_diskStream.get();
+	sf_close(m_sndfile);
+	delete audioBusHandle();
+}
+
+void SamplePreviewPlayHandle::play(SampleFrame* dst)
+{
+	// TODO: Should be a parameter in the function
+	const auto dstFrameCount = static_cast<sf_count_t>(Engine::audioEngine()->framesPerPeriod());
+
+	const auto srcFramesNeeded = static_cast<sf_count_t>(std::ceil(dstFrameCount / resamplingRatio()));
+	const auto srcFramesToRead = std::min<sf_count_t>(framesAvailableToRead(), srcFramesNeeded);
+	if (srcFramesToRead == 0) { return; }
+
+	const auto channelsToRead = std::min(m_sfinfo.channels, static_cast<int>(DEFAULT_CHANNELS));
+	const auto frameReadIndex = m_frameReadIndex.load(std::memory_order::acquire);
+
+	for (auto dstFrameIndex = 0; dstFrameIndex < dstFrameCount; ++dstFrameIndex)
+	{
+		const auto srcFrameIndex = dstFrameIndex / resamplingRatio();
+
+		if (srcFrameIndex >= static_cast<double>(srcFramesToRead))
+		{
+			dst[dstFrameIndex][0] = 0.f;
+			dst[dstFrameIndex][1] = 0.f;
+			continue;
+		}
+
+		const auto currentFrameReadIndex = static_cast<double>(frameReadIndex) + srcFrameIndex;
+		const auto prevFrameReadIndex = std::floor(currentFrameReadIndex);
+		const auto nextFrameReadIndex = std::ceil(currentFrameReadIndex);
+
+		const auto prevSrcFrame = bufferAt(static_cast<sf_count_t>(prevFrameReadIndex));
+		const auto nextSrcFrame = bufferAt(static_cast<sf_count_t>(nextFrameReadIndex));
+
+		const auto interpolatedSrcFramePos = currentFrameReadIndex - prevFrameReadIndex;
+
+		for (auto channel = 0; channel < channelsToRead; ++channel)
+		{
+			const auto prevSrcSample = static_cast<double>(prevSrcFrame[channel]);
+			const auto nextSrcSample = static_cast<double>(nextSrcFrame[channel]);
+			const auto interpolatedSample = std::lerp(prevSrcSample, nextSrcSample, interpolatedSrcFramePos);
+
+			if (channelsToRead == 1)
+			{
+				dst[dstFrameIndex][0] = static_cast<float>(interpolatedSample);
+				dst[dstFrameIndex][1] = static_cast<float>(interpolatedSample);
+			}
+			else { dst[dstFrameIndex][channel] = static_cast<float>(interpolatedSample); }
+		}
```

Comment:
```suggestion
			dst[dstFrameIndex][channel] = static_cast<float>(interpolatedSample);
		}

		if (channelsToRead == 1)
		{
			dst[dstFrameIndex][1] = dst[dstFrameIndex][0];
		}
```
Simplification

Reply:
It's been awhile since I've looked at this code, but from what you've provided, aren't we saying that each frame only had one sample? Why are we indexing the frame by 1?

Reply:
The "1" is the channel, not the frame index.

I just moved the mono/stereo branching out of the inner for loop. Now the inner for loop is written the same way regardless of the number of channels, and afterwards the right channel's sample is copied from the left channel's when there is only one channel.

Reply:
Oh I see now, I think I read it wrong at first.